### DISCUSSION TITLE
Query Many-to-Many Relationships

### DISCUSSION THREAD

**Discussion by johannesschobel** [prisma/prisma#5678] [Q&A] [answered]
Dear all,

i would like to get the "best" (i.e., most suitable approach) for querying many-to-many relationships.

Consider the following example... i have `User` and `Course` models, whereas a `User` can participate in many `Course`s, and a `Course` has many `User`s that participate. i.e., there is a `many-to-many` association between those 2 relationships.

So far so good, i have set up those Prisma models accordingly and migrated to the database. The relationship table is created accordingly.

Everything fine :heavy_check_mark: 

Now, i have a route, where i would like to return all `Course`s of the currently logged in `User`.
What is the best approach to "query" through this n:m relationship?

Like so?
```ts
const userWithCourses = await this.prisma.user.findUnique({
  where: {id: userId},
  include: {courses: true}
});
return userWithCourses.courses;
```

or like this?
```ts
const courses = await this.prisma.courses.findMany({
   where: {participants: {some: {id: userId}}}
});
return courses;
```

Actually, i have a hard time wrapping my head around the `some`, `every` and `none` keywords in the `where.participants` clause.. i believe that the `some` keyword should get me the correct answer, but not sure about that, haha :laughing: 

All the best

**Comment by nikolasburk** [comment, marked as answer]
Hey Johannes, that's a great question! Ultimately I think you're right and there's no difference in the result of the two Prisma Client queries since they both express the same requirement. However, the SQL queries that are generated are different.

I just tested both queries in an isolated fashion and here's the result for both queries.

### `user.findUnique` with `include`

```ts
const userWithCourses = await this.prisma.user.findUnique({
  where: {id: userId},
  include: {courses: true}
});
return userWithCourses.courses;
```

This query produces the following SQL:

```sql
SELECT `dev`.`User`.`id`, `dev`.`User`.`name` FROM `dev`.`User` WHERE `dev`.`User`.`id` = ? LIMIT ? OFFSET ?
SELECT `dev`.`_CourseToUser`.`B`, `dev`.`_CourseToUser`.`A` FROM `dev`.`_CourseToUser` WHERE `dev`.`_CourseToUser`.`B` IN (?)
SELECT `dev`.`Course`.`id`, `dev`.`Course`.`title` FROM `dev`.`Course` WHERE `dev`.`Course`.`id` IN (?,?) LIMIT ? OFFSET ?
```

### `course.findMany` with `some`

```ts
const courses = await this.prisma.courses.findMany({
   where: {participants: {some: {id: userId}}}
});
return courses;
```

This query produces the following SQL:

```sql
SELECT `dev`.`Course`.`id`, `dev`.`Course`.`title` FROM `dev`.`Course` WHERE (`dev`.`Course`.`id`) IN (SELECT `t0`.`A` FROM `dev`.`_CourseToUser` AS `t0` INNER JOIN `dev`.`User` AS `j0` ON (`j0`.`id`) = (`t0`.`B`) WHERE (`j0`.`id` = ? AND `t0`.`A` IS NOT NULL)) LIMIT ? OFFSET ?
````

### Conclusion

> What is the best approach to "query" through this n:m relationship?

I guess this depends on what exactly you're trying to achieve. With a very small dataset, the second query was slightly more performant than the first one on my machine.

Hope this helps, let me know if you have any further questions! 

(By the way, `some`, `none` and `every` are documented [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#relation-filters) and [here](https://www.prisma.io/docs/support/help-articles/finding-entities-based-on-relation) if you're looking for more explanations for how they works 🙂 )

**Reply by johannesschobel** [reply]
Dear @nikolasburk , 
thank you very much for your proper answer on this topic.
One additional benefit of approach 2 would be, that it is only one particular SQL statement that is executed. In terms of serializability and transactions this would be easier to handle.

All the best and thank you very much for your answer on this one.

Maybe other community members have additional thoughts and insights?

**Reply by lucascoelhosi** [reply]
Hi, commenting here to let you know that this solution helped me solve the many-to-many query in Prisma. The ORM is wonderful but it has some complicated things hehehe

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
